mm/pmm.c

```diff
@@ -117,13 +149,19 @@ static frames_array_t *new_frames_array(void) {
 
     init_frames_array(array);
 
-    total_free_frames += array->meta.free_count;
+    /* since paging will back off from refilling we are responsible to do it here */
+    try_create_4k_frames();
```

Comment:
In this approach, if paging detects that refilling is already active, it will not recreate the minimum 4k frames itself but delegate it to whoever is already refilling. This avoids double refilling of array frames and also avoids inconsistencies due to two interleaved array refills.